hadoop-yarn-project/hadoop-yarn/hadoop-yarn-client/src/test/java/org/apache/hadoop/yarn/client/api/impl/TestNMClient.java

```diff
@@ -125,576 +117,362 @@ public void postTransition(
+        assertEquals(0, nmClient.startedContainers.size());
+        return;
+      } catch (EarlyFinishException e) {
+        --earlyFinishCounter;
+      } finally {
+        tearDown();
+      }
+    }
+    if (earlyFinishCounter == 0) {
+      fail("Too many early finish exception happened");
+    }
   }
 
-  @Test (timeout = 200000)
+  @Test (timeout = 200_000 * MAX_EARLY_FINISH)
   public void testNMClient()
-      throws YarnException, IOException {
-    rmClient.registerApplicationMaster("Host", 10000, "");
-
-    testContainerManagement(nmClient, allocateContainers(rmClient, 5));
-    
-    rmClient.unregisterApplicationMaster(FinalApplicationStatus.SUCCEEDED,
-        null, null);
-    // stop the running containers on close
-    assertFalse(nmClient.startedContainers.isEmpty());
-    nmClient.cleanupRunningContainersOnStop(true);
+      throws YarnException, IOException, InterruptedException, TimeoutException {
+    int earlyFinishCounter = MAX_EARLY_FINISH;
+    while (0 < earlyFinishCounter) {
+      try {
+        setup();
+        rmClient.registerApplicationMaster("Host", 10_000, "");
+        testContainerManagement(nmClient, allocateContainers(rmClient));
+        rmClient.unregisterApplicationMaster(FinalApplicationStatus.SUCCEEDED, null, null);
+        // stop the running containers on close
+        assertFalse(nmClient.startedContainers.isEmpty());
+        nmClient.cleanupRunningContainersOnStop(true);
+        assertTrue(nmClient.getCleanupRunningContainers().get());
+        nmClient.stop();
+        return;
+      } catch (EarlyFinishException e) {
+        --earlyFinishCounter;
+      } finally {
+        tearDown();
+      }
+    }
+    if (earlyFinishCounter == 0) {
+      fail("Too many early finish exception happened");
+    }
+  }
+
+  private void stopNmClient() {
+    assertNotNull("Null nmClient", nmClient);
+    // leave one unclosed
+    assertEquals(1, nmClient.startedContainers.size());
+    // default true
     assertTrue(nmClient.getCleanupRunningContainers().get());
+    nmClient.cleanupRunningContainersOnStop(false);
+    assertFalse(nmClient.getCleanupRunningContainers().get());
     nmClient.stop();
   }
 
   private Set<Container> allocateContainers(
-      AMRMClientImpl<ContainerRequest> rmClient, int num)
-      throws YarnException, IOException {
-    // setup container request
-    Resource capability = Resource.newInstance(1024, 0);
-    Priority priority = Priority.newInstance(0);
-    String node = nodeReports.get(0).getNodeId().getHost();
-    String rack = nodeReports.get(0).getRackName();
-    String[] nodes = new String[] {node};
-    String[] racks = new String[] {rack};
-
-    for (int i = 0; i < num; ++i) {
-      rmClient.addContainerRequest(new ContainerRequest(capability, nodes,
-          racks, priority));
+      AMRMClientImpl<ContainerRequest> rmClient
+  ) throws YarnException, IOException {
+    for (int i = 0; i < NUMBER_OF_CONTAINERS; ++i) {
```

Comment:
Sure,
In short: this test is a mess
Long: I dig down deep into the code, and realised that code what you wrote always return with 5, so after that i took a shower, and purged it, and replaced with const 5

Reply:
OK. My point is: How I usually do refactor is to do it in steps:
First, extract methods + do some trivial refactors.
If I want to change the test execution logic, it can be done gradually in separate commits, not just a "big bang commit".
This way, separate fixes / changes can be reverted if they break something.